modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java

```diff
@@ -0,0 +1,275 @@
+    if (targetFlavorOption == null)
+      targetFlavorOption = StringUtils.trimToNull(operation.getConfiguration("target-flavors"));
+    String targetTagsOption = StringUtils.trimToNull(operation.getConfiguration("target-tags"));
+    String encodingProfileOption = StringUtils.trimToNull(operation.getConfiguration("encoding-profile"));
+    if (encodingProfileOption == null)
+      encodingProfileOption = StringUtils.trimToNull(operation.getConfiguration("encoding-profiles"));
+    String tagsAndFlavorsOption = StringUtils.trimToNull(operation.getConfiguration("tags-and-flavors"));
+    boolean tagsAndFlavors = BooleanUtils.toBoolean(tagsAndFlavorsOption);
+
+
+    MediaPackage mediaPackage = workflowInstance.getMediaPackage();
+
+    // Make sure either one of tags or flavors are provided
+    if (StringUtils.isBlank(sourceFlavorOption) && StringUtils.isBlank(sourceFlavorsOption)
+            && StringUtils.isBlank(sourceTagsOption)) {
+      logger.info("No source tags or flavors have been specified, not matching anything");
+      return createResult(mediaPackage, WorkflowOperationResult.Action.CONTINUE);
+    }
+
+    // Target flavor
+    MediaPackageElementFlavor targetFlavor = null;
+    if (StringUtils.isNotBlank(targetFlavorOption)) {
+      try {
+        targetFlavor = MediaPackageElementFlavor.parseFlavor(targetFlavorOption);
+      } catch (IllegalArgumentException e) {
+        throw new WorkflowOperationException("Target flavor '" + targetFlavorOption + "' is malformed");
+      }
+    }
+
+    List<String> profiles = new ArrayList<>();
+    for (String encodingProfileId : asList(encodingProfileOption)) {
+      EncodingProfile profile = composerService.getProfile(encodingProfileId);
+      if (profile == null)
+        throw new WorkflowOperationException("Encoding profile '" + encodingProfileId + "' was not found");
+      // just test if the profile exists, we only need the profile id for further work
+      profiles.add(encodingProfileId);
+    }
+
+    // Make sure there is at least one profile
+    if (profiles.isEmpty())
+      throw new WorkflowOperationException("No encoding profile was specified");
+
+    AttachmentSelector attachmentSelector = new AttachmentSelector();
+    for (String sourceFlavor : asList(sourceFlavorsOption)) {
+      attachmentSelector.addFlavor(sourceFlavor);
+    }
+    for (String sourceFlavor : asList(sourceFlavorOption)) {
+      attachmentSelector.addFlavor(sourceFlavor);
+    }
+    for (String sourceTag : asList(sourceTagsOption)) {
+      attachmentSelector.addTag(sourceTag);
+    }
+
+    // Look for elements matching the tag
+    Collection<Attachment> sourceElements = attachmentSelector.select(mediaPackage, tagsAndFlavors);
+
+    Map<Job, Attachment> jobs = new Hashtable<>();
+    try {
+      for (Attachment sourceElement : sourceElements) {
+        Job job = composerService.convertImage(sourceElement, profiles.toArray(new String[profiles.size()]));
```

Comment:
You are always converting lists to arrays and back. Why not always use a list?

Reply:
Because the API must be changed otherwise. I would unlikely do that.